### PR TITLE
Adding a feature to prometheusexporter to export exemplars along with…

### PR DESCRIPTION
**Pull request by arun-shopify**
**Description:** 
Adding a feature to `prometheusexporter` to export exemplars along with histogram metrics. In order to add this feature to the `prometheusexporter`, we made a PR (https://github.com/prometheus/client_golang/pull/986, merged) to add exemplars support to the prometheus/client_golang library.

Exemplars only show up in Open Metrics format, so I have added a config option `EnableOpenMetrics` in prometheus request handler that can be used to toggle the output format. 

**Testing:** 

Example config for testing the `prometheusexporter` in `openetelemerty-collector`:
 
```
receivers:
  nop: {}
  otlp:
    protocols:
      grpc:
      http:

processors:
  batch:
  spanmetrics:
    metrics_exporter: prometheus

exporters:
  nop: {}
  prometheus:
    endpoint: "localhost:8889"
    namespace: "test"
    enable_open_metrics: true

service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [spanmetrics]
      exporters: [nop]
    metrics:
      receivers: [nop]
      exporters: [prometheus]
```

To view the exemplars from prometheus exporter, you have to set the following header in your request.

`Accept: application/openmetrics-text;`

Example curl command:

```
curl -H 'Accept: application/openmetrics-text; version=0.0.1' http://localhost:8889/metrics
```

Sample output for one of the buckets:

```
test_latency_bucket{operation="TestSubSpan",service_name="tracegen",span_kind="SPAN_KIND_INTERNAL",status_code="STATUS_CODE_UNSET",le="4.0"} 140 1.645646235482e+09 # {trace_id="739d03769cc2b850576987fcec20c058"} 0.000583 1.645646235482858e+09
```

### CODE REVIEW

File: exporter/prometheusexporter/go.mod
Comment (mx-psi) [inline exporter/prometheusexporter/go.mod:15]: This module is deprecated and must not be used, you should use `go.opentelemetry.io/collector/pdata` or `go.opentelemetry.io/collector/semconv` instead

**Review by mx-psi — DISMISSED** [review]

**Comment by arun-shopify** [comment]
rebased this branch to the latest main.

File: exporter/prometheusexporter/go.mod
Comment (arun-shopify) [reply exporter/prometheusexporter/go.mod]: removed the use of `go.opentelemetry.io/collector/model`

**Comment by mx-psi** [comment]
@Aneurysm9 to review as a codeowner

**Comment by arun-shopify** [comment]
rebased this branch to the latest main branch and resolved merge conflicts.

**Comment by arun-shopify** [comment]
@Aneurysm9 would you please review this PR?

**Comment by arun-shopify** [comment]
@mx-psi @Aneurysm9 bumping this PR for review again :)

**Comment by arun-shopify** [comment]
@mx-psi fixed that CI error, could you please rerun CI to verify? Also, I rebased to the latest main branch.

**Comment by mx-psi** [comment]
@arun-shopify I ran the tests again, it looks like the build is still failing. You should be able to reproduce it locally by running `make otelcontribcol'.

**Comment by arun-shopify** [comment]
> `make otelcontribcol'

thanks @mx-psi , that issue should be fixed now.

File: go.mod
Comment (mx-psi) [inline go.mod:874]: Why do we need this? Is this needed on every module that imports the prometheus exporter module?

Comment (arun-shopify) [reply go.mod:874]: not every module but some tests like `cmd/otelcontribcol/main_testbed.go` fails (because it runs prometheusexporter) and there is no `go.mod` in that directory to include it in there so I put it in the main go.mod

Comment (mx-psi) [reply go.mod:874]: 🤔 I would expect that removing the `replace` and running `make gotidy` would fix this.

Comment (mx-psi) [reply go.mod:874]: I will have a look in the next few days to see if I can make it work. Using a replace probably means that downstream distributions will need to use this too, so if we can avoid it it would be great

Comment (mx-psi) [reply go.mod:874]: Hey, sorry for the delay. I have tried using the latest commit, but it seems like Go considers v1.12.2 to be higher than the latest commit and decides to use this. I am happy to approve this once 
1. the test suite passes and
2. you document why the `replace` is needed, so that downstream distros can understand what they need to do to accomodate this.

Comment (arun-shopify) [reply go.mod:906]: Rebased to the latest main. Added a comment in go.mod explaining why `replace` is needed, please let me know if that's all that is needed there. Fixed couple more CI failures. Could you please rerun the CI?

File: exporter/prometheusexporter/config.go
Comment (mx-psi) [inline exporter/prometheusexporter/config.go:50]: Wondering if this should be inside a section for handler options, so that we can namespace this

**Review by mx-psi — COMMENTED** [review]
Thanks for addressing the changes! I think you will need a changelog entry for this to pass. I also left a small (non-blocking) comment to know your opinion

**Comment by mx-psi** [comment]
Can you merge main to your branch? Tests were accidentally broken by a PR and are not running. Sorry that it's taking so long for this PR to be merged, if tests pass after that everything looks good!

**Comment by mx-psi** [comment]
Ci is failing in the linter and the go mod tidy check

**Comment by arun-shopify** [comment]
fixed those; could you please rerun the CI @mx-psi ? thanks

File: exporter/prometheusexporter/config.go
Comment (arun-shopify) [reply exporter/prometheusexporter/config.go:49]: I thought requiring a different export format per namespace would be a rather rare use case and so making this an exporter level config made more sense.

**Comment by arun-shopify** [comment]
looks like the CI passed this time! I just pushed some read me update.

**Review by mx-psi — APPROVED** [review]

**Comment by mx-psi** [comment]
✅, but it looks like there is a conflict (I can't see what exactly, Github doesn't show it 🤔). Could you fix it?

**Comment by arun-shopify** [comment]
there is new conflict in CHANGELOG.md, just from a commit 1 minute ago :)

**Comment by alvinhom** [comment]
Hi, I was waiting to test the exemplar support from the prometheus exporter.  So, I did a local build with this PR.  When I restarted my collector, I am seeing the following error in the collector logs:

2022/06/28 16:48:50 http: panic serving 10.42.15.30:39238: no bucket was found for given exemplar value
goroutine 184 [running]:
net/http.(*conn).serve.func1()
        net/http/server.go:1825 +0xbf
panic({0x453d2e0, 0x5da0920})
        runtime/panic.go:844 +0x258
github.com/prometheus/client_golang/prometheus.(*withExemplarsMetric).Write(0xc001504570, 0xc000234770)
        github.com/prometheus/client_golang@v1.12.2/prometheus/metric.go:188 +0x1c5
github.com/prometheus/client_golang/prometheus.processMetric({0x5dc3a98, 0xc001504570}, 0xc00098e180?, 0x0?, 0x0)
        github.com/prometheus/client_golang@v1.12.2/prometheus/registry.go:605 +0x98
github.com/prometheus/client_golang/prometheus.(*Registry).Gather(0xc0009b3220)
        github.com/prometheus/client_golang@v1.12.2/prometheus/registry.go:509 +0x71f
github.com/prometheus/client_golang/prometheus.(*noTransactionGatherer).Gather(0x8?)
        github.com/prometheus/client_golang@v1.12.2/prometheus/registry.go:1041 +0x22
github.com/prometheus/client_golang/prometheus/promhttp.HandlerForTransactional.func1({0x5de35c8, 0xc0013f08c0}, 0xc001c4a700)
        github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/http.go:133 +0xfe
net/http.HandlerFunc.ServeHTTP(0x7f1fcb879fc0?, {0x5de35c8?, 0xc0013f08c0?}, 0x40da65?)
        net/http/server.go:2084 +0x2f
net/http.(*ServeMux).ServeHTTP(0x0?, {0x5de35c8, 0xc0013f08c0}, 0xc001c4a700)
        net/http/server.go:2462 +0x149
net/http.serverHandler.ServeHTTP({0xc0013b5d10?}, {0x5de35c8, 0xc0013f08c0}, 0xc001c4a700)
        net/http/server.go:2916 +0x43b
net/http.(*conn).serve(0xc000475400, {0x5de5d88, 0xc0004553b0})
        net/http/server.go:1966 +0x5d7
created by net/http.(*Server).Serve
        net/http/server.go:3071 +0x4db

If I curl the metrics endpoint, it returns with empty data.  My pipeline goes like this:

Java Agent (otel exporter trace/metrics)  -> otel collector (prometheus exporter) -> prometheus

The metrics was working fine previously, however it was not getting exemplar obviously.  

I tried enable and disabling the "enable_open_metrics" flag, but it does not seem to matter and I get the same error.

**Comment by arun-shopify** [comment]
> Hi, I was waiting to test the exemplar support from the prometheus exporter. So, I did a local build with this PR. When I restarted my collector, I am seeing the following error in the collector logs:
> 
> 2022/06/28 16:48:50 http: panic serving 10.42.15.30:39238: no bucket was found for given exemplar value goroutine 184 [running]: net/http.(*conn).serve.func1() net/http/server.go:1825 +0xbf panic({0x453d2e0, 0x5da0920}) runtime/panic.go:844 +0x258 github.com/prometheus/client_golang/prometheus.(*withExemplarsMetric).Write(0xc001504570, 0xc000234770) github.com/prometheus/client_golang@v1.12.2/prometheus/metric.go:188 +0x1c5 github.com/prometheus/client_golang/prometheus.processMetric({0x5dc3a98, 0xc001504570}, 0xc00098e180?, 0x0?, 0x0) github.com/prometheus/client_golang@v1.12.2/prometheus/registry.go:605 +0x98 github.com/prometheus/client_golang/prometheus.(*Registry).Gather(0xc0009b3220) github.com/prometheus/client_golang@v1.12.2/prometheus/registry.go:509 +0x71f github.com/prometheus/client_golang/prometheus.(*noTransactionGatherer).Gather(0x8?) github.com/prometheus/client_golang@v1.12.2/prometheus/registry.go:1041 +0x22 github.com/prometheus/client_golang/prometheus/promhttp.HandlerForTransactional.func1({0x5de35c8, 0xc0013f08c0}, 0xc001c4a700) github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/http.go:133 +0xfe net/http.HandlerFunc.ServeHTTP(0x7f1fcb879fc0?, {0x5de35c8?, 0xc0013f08c0?}, 0x40da65?) net/http/server.go:2084 +0x2f net/http.(*ServeMux).ServeHTTP(0x0?, {0x5de35c8, 0xc0013f08c0}, 0xc001c4a700) net/http/server.go:2462 +0x149 net/http.serverHandler.ServeHTTP({0xc0013b5d10?}, {0x5de35c8, 0xc0013f08c0}, 0xc001c4a700) net/http/server.go:2916 +0x43b net/http.(*conn).serve(0xc000475400, {0x5de5d88, 0xc0004553b0}) net/http/server.go:1966 +0x5d7 created by net/http.(*Server).Serve net/http/server.go:3071 +0x4db
> 
> If I curl the metrics endpoint, it returns with empty data. My pipeline goes like this:
> 
> Java Agent (otel exporter trace/metrics) -> otel collector (prometheus exporter) -> prometheus
> 
> The metrics was working fine previously, however it was not getting exemplar obviously.
> 
> I tried enable and disabling the "enable_open_metrics" flag, but it does not seem to matter and I get the same error.

@alvinhom, I tested it out using the latest `prometheusexporter` code from the Otel contrib repo with the latest Otel collector (v0.54.0) and it seemed to work just fine. Here is the config that I used for testing:
```
receivers:
  # Dummy receiver that's never used, because a pipeline is required to have one.
  otlp/spanmetrics:
    protocols:
      grpc:
        endpoint: "localhost:12345"
  otlp:
    protocols:
      grpc:
      http:

processors:
  batch:
  spanmetrics:
    metrics_exporter: prometheus
    dimensions_cache_size: 500

exporters:
  logging:
    logLevel: debug
  prometheus:
    endpoint: "localhost:8889"
    send_timestamps: true
    namespace: "test"
    enable_open_metrics: true

service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [spanmetrics]
      exporters: [logging]
    metrics:
      receivers: [otlp/spanmetrics]
      exporters: [prometheus]
```

**Comment by mx-psi** [comment]
@alvinhom if the issue is not (necessarily) related to this PR, I would suggest you open an issue for this, so that more people can easily see it and participate

**Comment by alvinhom** [comment]
I think the issue is if there are exemplars associated with the last bucket x:Inf.

I added the following code to the collector_test.go:TestConvertDoubleHistogramExemplar

	exemplars := []prometheus.Exemplar{
		{
			Timestamp: exemplarTs,
			Value:     3,
			Labels:    prometheus.Labels{"test_label_0": "label_value_0"},
		},
		{
			Timestamp: exemplarTs,
			Value:     50,
			Labels:    prometheus.Labels{"test_label_1": "label_value_1"},
		},
		{
			Timestamp: exemplarTs,
			Value:     78,
			Labels:    prometheus.Labels{"test_label_2": "label_value_2"},
		},
		**{
			Timestamp: exemplarTs,
			Value:     100,
			Labels:    prometheus.Labels{"test_label_3": "label_value_3"},
		},**
	}

	require.Equal(t, 100.0, buckets[3].GetExemplar().GetValue())
	require.Equal(t, int32(128654848), buckets[3].GetExemplar().GetTimestamp().GetNanos())
	require.Equal(t, 1, len(buckets[3].GetExemplar().GetLabel()))
	require.Equal(t, "test_label_3", buckets[3].GetExemplar().GetLabel()[0].GetName())
	require.Equal(t, "label_value_3", buckets[3].GetExemplar().GetLabel()[0].GetValue())

Basically, adding an exemplar associated with the last bucket.

When I tried to run the test, I get the same errors:

--- FAIL: TestConvertDoubleHistogramExemplar (0.00s)
panic: no bucket was found for given exemplar value [recovered]
        panic: no bucket was found for given exemplar value

goroutine 148 [running]:
testing.tRunner.func1.2({0x4597040, 0x5dc9da0})
        /usr/local/Cellar/go/1.18.3/libexec/src/testing/testing.go:1389 +0x366
testing.tRunner.func1()
        /usr/local/Cellar/go/1.18.3/libexec/src/testing/testing.go:1392 +0x5d2
panic({0x4597040, 0x5dc9da0})
        /usr/local/Cellar/go/1.18.3/libexec/src/runtime/panic.go:844 +0x258
github.com/prometheus/client_golang/prometheus.(*withExemplarsMetric).Write(0xc000492a50, 0xc000a93ab0)
        /Users/ahom/Downloads/dev/go/pkg/mod/github.com/prometheus/client_golang@v1.12.2-0.20220318110013-3bc8f2c651ff/prometheus/metric.go:188 +0x465
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusexporter.TestConvertDoubleHistogramExemplar(0x0?)
        /Users/ahom/Downloads/dev/linux/opentelemetry-collector-contrib-snapshot/exporter/prometheusexporter/collector_test.go:161 +0x11c8
testing.tRunner(0xc0006031e0, 0x4eab610)
        /usr/local/Cellar/go/1.18.3/libexec/src/testing/testing.go:1439 +0x214
created by testing.(*T).Run
        /usr/local/Cellar/go/1.18.3/libexec/src/testing/testing.go:1486 +0x725
FAIL    github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusexporter   1.354s
FAIL
make[2]: *** [test] Error 1

**Comment by mx-psi** [comment]
So indeed the panic only happens after this PR. From the comment above the line where this panics:
https://github.com/prometheus/client_golang/blob/66837e3298bdc57a828794c23bacb253ed8c04cd/prometheus/metric.go#L187-L188

I would assume this is not supported and that the problem comes from the application producing the data, but it would be great to confirm this

**Comment by arun-shopify** [comment]
support for Inf bucket was removed in the latest `prometheus/client_golang` release, so thats why the panic: https://github.com/prometheus/client_golang/releases/tag/v1.12.2

**Comment by arun-shopify** [comment]
So the panic is because the value found does not belong to any bucket bounds. @alvinhom would the value happen be outside of the default bucket bound `DefBuckets = []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10}` ?

**Comment by arun-shopify** [comment]
@alvinhom in your test case, if the value is `89` it works since it is less than the highest upper bound in: `[]float64{5, 25, 90}`

**Comment by mx-psi** [comment]
@arun-shopify @alvinhom after discussion with the release manager, since this is a widely used component and the issue can cause the Collector to crash without any known workaround, we have decided to revert this PR before it lands in any Collector release (the revert PR is #12074). 

The functionality added by this PR can of course be accepted in a future PR, after we understand the issue better.

**Comment by arun-shopify** [comment]
@alvinhom would you be able share the config you are using?

**Comment by alvinhom** [comment]
> So indeed the panic only happens after this PR. From the comment above the line where this panics: https://github.com/prometheus/client_golang/blob/66837e3298bdc57a828794c23bacb253ed8c04cd/prometheus/metric.go#L187-L188
> 
> I would assume this is not supported and that the problem comes from the application producing the data, but it would be great to confirm this

The metric data is produced by the Java opentelemetry agent.  The agent just outputs the data and forwards to the otel-collector, and I use prometheus exporter to export that to Prometheus.I do not have any custom instrumentation code.  

Also, I would assume that exemplar data from the largest boundary to inf should be supported.  I am not sure why it is not supported.

**Comment by alvinhom** [comment]
> @alvinhom would you be able share the config you are using?

The data pipeline is like this:

OpenTelemetry Java Instrumentation agent (with auto instrumentation) -->  OTel Collector Contrib --> Prometheus

The collector configuration file is like this:

    receivers:
      otlp:
        protocols:
          grpc:
    processors:
      tail_sampling:
        decision_wait: 5s
        policies:
          [
            {
               name: probability,
               type: probabilistic,
               probabilistic: { sampling_percentage: 10 },
            },
            { name: error,
              type: status_code,
              status_code: { status_codes: [ERROR]} },
            { name: healthcheck,
              type: string_attribute,
              string_attribute: {
                   key: 'http.target',
                   values: ['^\/(?:metrics\/.*|.*readiness|.*liveness|.*version|.*health|.*version.jsp|.*health.jsp)$'],
                   enabled_regex_matching: true,
                   invert_match: true } },
            { name: slow,
              type: latency,
              latency: { threshold_ms: 1000 } },
          ]
    exporters:
      prometheus:
        endpoint: "0.0.0.0:8889"
        resource_to_telemetry_conversion:
          enabled: true
        enable_open_metrics: true
      jaeger:
        endpoint: "jaeger-es-collector.observability.svc:14250"
        tls:
          insecure: true
    service:
      pipelines:
        metrics:
          receivers: [otlp]
          exporters: [prometheus]
        traces:
          receivers: [otlp]
          processors: [tail_sampling]
          exporters: [jaeger]

**Comment by alvinhom** [comment]
I implemented the workaround as suggested by above to filter out exemplars at the +Inf bucket and the panic is gone.  The fix should really be in the prometheus/client_golang.  

I have also found that with my pipeline from Java agent -> otel collector, the trace id is not set in the prometheus labels.  I had to add code to get the traceid and spanid and add it to the Prometheus labels.

With the above changes, I was able to get the exemplars working and integrate it with Grafana and see the exemplars.

**Comment by arun-shopify** [comment]
new PR with the fix:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13127

**Pull request merged**
